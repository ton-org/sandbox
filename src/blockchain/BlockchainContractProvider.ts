import {
    AccountState,
    Address,
    Cell,
    comment,
    Contract,
    ContractGetMethodResult,
    ContractProvider,
    ContractState,
    ExtraCurrency,
    Message,
    OpenedContract,
    Sender,
    SendMode,
    StateInit,
    toNano,
    Transaction,
    TupleItem,
} from '@ton/core';

import { TickOrTock } from '../executor/Executor';
import { GetMethodResult, SmartContract } from './SmartContract';

function bigintToBuffer(x: bigint, n = 32): Buffer {
    const b = Buffer.alloc(n);
    for (let i = 0; i < n; i++) {
        b[n - i - 1] = Number((x >> BigInt(i * 8)) & 0xffn);
    }
    return b;
}

function convertState(state: AccountState | undefined): ContractState['state'] {
    if (state === undefined)
        return {
            type: 'uninit',
        };

    switch (state.type) {
        case 'uninit':
            return {
                type: 'uninit',
            };
        case 'active':
            return {
                type: 'active',
                code: state.state.code?.toBoc(),
                data: state.state.data?.toBoc(),
            };
        case 'frozen':
            return {
                type: 'frozen',
                stateHash: bigintToBuffer(state.stateHash),
            };
    }
}

export interface SandboxContractProvider extends ContractProvider {
    tickTock(which: TickOrTock): Promise<void>;
}

/**
 * Provider used in contracts to send messages or invoke getters. For additional information see {@link Blockchain.provider}
 */
export class BlockchainContractProvider implements SandboxContractProvider {
    constructor(
        private readonly blockchain: {
            getContract(address: Address): Promise<SmartContract>;
            pushMessage(message: Message): Promise<void>;
            runGetMethod(address: Address, method: string, args: TupleItem[]): Promise<GetMethodResult>;
            pushTickTock(on: Address, which: TickOrTock): Promise<void>;
            openContract<T extends Contract>(contract: T): OpenedContract<T>;
        },
        private readonly address: Address,
        private readonly init?: StateInit | null,
    ) {}

    /**
     * Opens contract. For additional information see {@link Blockchain.open}
     */
    open<T extends Contract>(contract: T): OpenedContract<T> {
        return this.blockchain.openContract(contract);
    }

    async getState(): Promise<ContractState> {
        const contract = await this.blockchain.getContract(this.address);
        return {
            balance: contract.balance,
            extracurrency: contract.ec,
            last: {
                lt: contract.lastTransactionLt,
                hash: bigintToBuffer(contract.lastTransactionHash),
            },
            state: convertState(contract.accountState),
        };
    }

    /**
     * Invokes get method.
     * @param name Name of get method
     * @param args Args to invoke get method.
     */
    async get(name: string, args: TupleItem[]): Promise<ContractGetMethodResult> {
        const result = await this.blockchain.runGetMethod(this.address, name, args);
        const ret = {
            ...result,
            stack: result.stackReader,
            stackItems: result.stack,
            logs: result.vmLogs,
        };
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        delete (ret as any).stackReader;
        return ret;
    }

    /**
     * Dummy implementation of getTransactions. Sandbox does not store transactions, so its ContractProvider cannot fetch any.
     * Throws error in every call.
     *
     * @throws {Error}
     */
    getTransactions(
        _address: Address,
        _lt: bigint,
        _hash: Buffer,
        _limit?: number | undefined,
    ): Promise<Transaction[]> {
        throw new Error(
            '`getTransactions` is not implemented in `BlockchainContractProvider`, do not use it in the tests',
        );
    }

    /**
     * Pushes external-in message to message queue.
     * @param message Message to push
     */
    async external(message: Cell) {
        const init = (await this.getState()).state.type !== 'active' && this.init ? this.init : undefined;

        await this.blockchain.pushMessage({
            info: {
                type: 'external-in',
                dest: this.address,
                importFee: 0n,
            },
            init,
            body: message,
        });
    }

    /**
     * Pushes internal message to message queue.
     */
    async internal(
        via: Sender,
        args: {
            value: string | bigint;
            extracurrency?: ExtraCurrency;
            bounce?: boolean | null;
            sendMode?: SendMode;
            body?: string | Cell | null;
        },
    ) {
        const init = (await this.getState()).state.type !== 'active' && this.init ? this.init : undefined;

        const bounce = args.bounce !== null && args.bounce !== undefined ? args.bounce : true;

        const value = typeof args.value === 'string' ? toNano(args.value) : args.value;

        const body = typeof args.body === 'string' ? comment(args.body) : args.body;

        await via.send({
            to: this.address,
            value,
            bounce,
            extracurrency: args.extracurrency,
            sendMode: args.sendMode,
            init,
            body,
        });
    }

    /**
     * Pushes tick-tock message to message queue.
     */
    async tickTock(which: TickOrTock) {
        await this.blockchain.pushTickTock(this.address, which);
    }
}
