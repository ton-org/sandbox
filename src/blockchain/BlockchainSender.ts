import { Address, Cell, Message, packExtraCurrencyDict, Sender, SenderArguments } from '@ton/core';

/**
 * Sender for sandbox blockchain. For additional information see {@link Blockchain.sender}
 */
export class BlockchainSender implements Sender {
    constructor(
        private readonly blockchain: {
            pushMessage(message: Message): Promise<void>;
        },
        readonly address: Address,
    ) {}

    async send(args: SenderArguments) {
        await this.blockchain.pushMessage({
            info: {
                type: 'internal',
                ihrDisabled: true,
                ihrFee: 0n,
                bounce: args.bounce ?? true,
                bounced: false,
                src: this.address,
                dest: args.to,
                value: {
                    coins: args.value,
                    other: args.extracurrency ? packExtraCurrencyDict(args.extracurrency) : undefined,
                },
                forwardFee: 0n,
                createdAt: 0,
                createdLt: 0n,
            },
            body: args.body ?? new Cell(),
        });
    }
}
