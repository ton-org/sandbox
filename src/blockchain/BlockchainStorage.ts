import {AccountState, Address, Cell} from "@ton/core";
import {SmartContract} from "./SmartContract";
import {Blockchain} from "./Blockchain";


/**
 * @interface BlockchainStorage Provides information about contracts by blockchain
 */
export interface BlockchainStorage {
    /**
     * Retrieves a smart contract by blockchain and address.
     *
     * @param {Blockchain} blockchain - The blockchain instance.
     * @param {Address} address - The address of the smart contract.
     * @returns {Promise<SmartContract>} - The smart contract instance.
     */
    getContract(blockchain: Blockchain, address: Address): Promise<SmartContract>
    /**
     * Lists all known smart contracts.
     *
     * @returns {SmartContract[]} - An array of known smart contracts.
     */
    knownContracts(): SmartContract[]

    /**
     * Clears the internal cache of known contracts.
     */
    clearKnownContracts(): void
}

/**
 * In-memory storage for blockchain smart contracts.
 */
export class LocalBlockchainStorage implements BlockchainStorage {
    private contracts: Map<string, SmartContract> = new Map()

    async getContract(blockchain: Blockchain, address: Address) {
        let existing = this.contracts.get(address.toString())
        if (!existing) {
            existing = SmartContract.empty(blockchain, address)
            this.contracts.set(address.toString(), existing)
        }

        return existing
    }

    knownContracts() {
        return Array.from(this.contracts.values())
    }

    clearKnownContracts() {
        this.contracts.clear()
    }
}

export interface RemoteBlockchainStorageClient {
    getLastBlockSeqno(): Promise<number>
    getAccount(seqno: number, address: Address): Promise<{
        state: AccountState
        balance: bigint
        lastTransaction?: { lt: bigint, hash: Buffer }
    }>
}

function convertTonClient4State(state: {
    type: "uninit";
} | {
    type: "active";
    code: string | null;
    data: string | null;
} | {
    type: "frozen";
    stateHash: string;
}): AccountState {
    switch (state.type) {
        case 'uninit':
            return { type: 'uninit' }
        case 'active':
            return { type: 'active', state: { code: state.code === null ? undefined : Cell.fromBase64(state.code), data: state.data === null ? undefined : Cell.fromBase64(state.data) } }
        case 'frozen':
            return { type: 'frozen', stateHash: BigInt('0x' + Buffer.from(state.stateHash, 'base64').toString('hex')) }
        default:
            throw new Error(`Unknown type ${state}`)
    }
}

/**
 * Wraps ton client for remote storage.
 *
 * @example
 * let client = new TonClient4({
 *     endpoint: 'https://mainnet-v4.tonhubapi.com'
 * })
 *
 * let remoteStorageClient = wrapTonClient4ForRemote(client);
 *
 * @param client TonClient4 to wrap
 */
export function wrapTonClient4ForRemote(client: {
    getLastBlock(): Promise<{
        last: {
            seqno: number;
        };
    }>;
    getAccount(seqno: number, address: Address): Promise<{
        account: {
            state: {
                type: "uninit";
            } | {
                type: "active";
                code: string | null;
                data: string | null;
            } | {
                type: "frozen";
                stateHash: string;
            };
            balance: {
                coins: string;
            };
            last: {
                lt: string;
                hash: string;
            } | null;
        };
    }>;
}): RemoteBlockchainStorageClient {
    return {
        getLastBlockSeqno: async () => {
            const last = await client.getLastBlock()
            return last.last.seqno
        },
        getAccount: async (seqno: number, address: Address) => {
            const { account } = await client.getAccount(seqno, address)
            return {
                state: convertTonClient4State(account.state),
                balance: BigInt(account.balance.coins),
                lastTransaction: account.last === null ? undefined : {
                    lt: BigInt(account.last.lt),
                    hash: Buffer.from(account.last.hash, 'base64'),
                },
            }
        },
    }
}

/**
 * @class {RemoteBlockchainStorage} Remote blockchain storage implementation.
 *
 * @example
 * let client = new TonClient4({
 *     endpoint: 'https://mainnet-v4.tonhubapi.com'
 * })
 *
 * let blockchain = await Blockchain.create({
 *     storage: new RemoteBlockchainStorage(wrapTonClient4ForRemote(client), 34892000)
 * });
 */
export class RemoteBlockchainStorage implements BlockchainStorage {
    private contracts: Map<string, SmartContract> = new Map()
    private client: RemoteBlockchainStorageClient
    private blockSeqno?: number

    constructor(client: RemoteBlockchainStorageClient, blockSeqno?: number) {
        this.client = client
        this.blockSeqno = blockSeqno
    }

    private async getLastBlockSeqno() {
        return this.blockSeqno ?? await this.client.getLastBlockSeqno()
    }

    async getContract(blockchain: Blockchain, address: Address) {
        let existing = this.contracts.get(address.toString())
        if (!existing) {
            let blockSeqno = await this.getLastBlockSeqno()
            let account = await this.client.getAccount(blockSeqno, address)

            const lt = account.lastTransaction?.lt ?? 0n

            existing = new SmartContract({
                lastTransactionHash: BigInt('0x' + (account.lastTransaction?.hash?.toString('hex') ?? '0')),
                lastTransactionLt: lt,
                account: {
                    addr: address,
                    storageStats: {
                        used: {
                            cells: 0n,
                            bits: 0n,
                            publicCells: 0n,
                        },
                        lastPaid: 0,
                        duePayment: null,
                    },
                    storage: {
                        lastTransLt: lt === 0n ? 0n : (lt + 1n),
                        balance: { coins: account.balance },
                        state: account.state,
                    },
                },
            }, blockchain)

            this.contracts.set(address.toString(), existing)
        }

        return existing
    }

    knownContracts() {
        return Array.from(this.contracts.values())
    }

    clearKnownContracts() {
        this.contracts.clear()
    }
}
