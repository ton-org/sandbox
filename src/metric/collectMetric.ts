import { Address, beginCell, Cell, Contract, storeMessage, Message } from '@ton/core';
import { Dictionary, DictionaryKeyTypes, TransactionComputePhase } from '@ton/core';
import { Maybe } from '@ton/core/src/utils/maybe';

import { Blockchain, SendMessageResult } from '../blockchain/Blockchain';
import { ContractDatabase } from './ContractDatabase';

export type MetricContext<T extends Contract> = {
    contract: T;
    methodName: string;
};

type StateShort = {
    code: Cell;
    data: Cell;
};

export type CellMetric = {
    cells: number;
    bits: number;
};

export type ComputePhaseMetric = {
    type: string;
    success?: boolean;
    gasUsed?: number;
    exitCode?: number;
    vmSteps?: number;
};

export type ActionPhaseMetric = {
    success: boolean;
    totalActions: number;
    skippedActions: number;
    resultCode: number;
    totalFwdFees?: number;
    totalActionFees: number;
    totalMessageSize: CellMetric;
};

export type AddressFriendly = string;

export function isAddressFriendly(value: unknown): value is AddressFriendly {
    return typeof value === 'string' && Address.isFriendly(value);
}

export type ContractName = string;

export type ContractMethodName = string;

export type CodeHash = `0x${string}`;

export type OpCode = `0x${string}`;

export function isCodeHash(value: unknown): value is CodeHash {
    return typeof value === 'string' && value.length === 66 && /^0x[0-9a-fA-F]+$/.test(value);
}

export type StateMetric = {
    code: CellMetric;
    data: CellMetric;
};

export type Metric = {
    // the name of the current test (if available in Jest context)
    testName?: string;
    // address of contract
    address: AddressFriendly;
    // hex-formatted hash of contract code
    codeHash?: CodeHash;
    // total cells and bits usage of the contract's code and data
    state: StateMetric;
    contractName?: ContractName;
    methodName?: ContractMethodName;
    receiver?: 'internal' | 'external-in' | 'external-out';
    opCode: OpCode;
    // information from transaction phases
    execute: {
        compute: ComputePhaseMetric;
        action?: ActionPhaseMetric;
    };
    // total cells and bits usage of inbound and outbound messages
    message: {
        in: CellMetric;
        out: CellMetric;
    };
};

export type SnapshotMetric = {
    label: string;
    createdAt: Date;
    items: Metric[];
};

interface HasCreatedAt {
    createdAt: Date;
}

export function sortByCreatedAt(reverse = false) {
    return (a: HasCreatedAt, b: HasCreatedAt) => (a.createdAt.getTime() - b.createdAt.getTime()) * (reverse ? -1 : 1);
}

export type SnapshotMetricFile = {
    name: string;
    content: SnapshotMetric;
};

export type SnapshotMetricList = Record<string, SnapshotMetricFile>;

export type SnapshotMetricConfig = {
    label: string;
    contractExcludes: ContractName[];
    contractDatabase: ContractDatabase;
};

const STORE_METRIC = Symbol.for('ton-sandbox-metric-store');

export function makeSnapshotMetric(store: Metric[], config: Partial<SnapshotMetricConfig> = {}): SnapshotMetric {
    const label = config.label || 'current';
    const contractExcludes = config.contractExcludes || new Array<ContractName>();
    const contractDatabase = config.contractDatabase || ContractDatabase.from({});
    const snapshot: SnapshotMetric = {
        label,
        createdAt: new Date(),
        items: new Array<Metric>(),
    };
    // remove duplicates and extract ABI
    const seen = new Set<string>();
    for (const metric of store) {
        const key = JSON.stringify(metric);
        if (seen.has(key)) continue;
        snapshot.items.push(metric);
        seen.add(key);
        if (metric.codeHash) {
            contractDatabase.extract(metric);
        }
    }
    // ABI auto-mapping
    for (const item of snapshot.items) {
        const find = contractDatabase.by(item);
        if (!item.contractName && find.contractName) {
            item.contractName = find.contractName;
        }
        if (!item.methodName && find.methodName) {
            item.methodName = find.methodName;
        }
    }
    if (contractExcludes.length > 0) {
        snapshot.items = snapshot.items.filter(
            (it) => typeof it.contractName === 'undefined' || !contractExcludes.includes(it.contractName),
        );
    }
    return snapshot;
}

export function getMetricStore(context: any = globalThis): Array<Metric> | undefined {
    return context[STORE_METRIC];
}

export function createMetricStore(context: any = globalThis): Array<Metric> {
    if (!Array.isArray(context[STORE_METRIC])) {
        context[STORE_METRIC] = new Array<Metric>();
    }
    return context[STORE_METRIC];
}

export function resetMetricStore(context: any = globalThis): Array<Metric> {
    const store = getMetricStore(context);
    if (store) store.length = 0;
    return createMetricStore(context);
}

export function calcMessageSize(msg: Maybe<Message>) {
    if (msg) {
        return calcCellSize(beginCell().store(storeMessage(msg)).endCell());
    }
    return { cells: 0, bits: 0 };
}

export function calcDictSize<K extends DictionaryKeyTypes, V>(dict: Dictionary<K, V>) {
    if (dict.size > 0) {
        return calcCellSize(beginCell().storeDict(dict).endCell().asSlice().loadRef());
    }
    return { cells: 0, bits: 0 };
}

export function calcCellSize(root: Cell, visited: Set<string> = new Set<string>()) {
    const hash = root.hash().toString('hex');
    if (visited.has(hash)) {
        return { cells: 0, bits: 0 };
    }
    visited.add(hash);
    const out = {
        cells: 1,
        bits: root.bits.length,
    };
    for (const ref of root.refs) {
        const childRes = calcCellSize(ref, visited);
        out.cells += childRes.cells;
        out.bits += childRes.bits;
    }
    return out;
}

export function calcStateSize(state: StateShort): StateMetric {
    const codeSize = calcCellSize(state.code);
    const dataSize = calcCellSize(state.data);
    return {
        code: codeSize,
        data: dataSize,
    };
}

export function calcComputePhase(phase: TransactionComputePhase): ComputePhaseMetric {
    if (phase.type === 'vm') {
        return {
            type: phase.type,
            success: phase.success,
            gasUsed: Number(phase.gasUsed),
            exitCode: phase.exitCode,
            vmSteps: phase.vmSteps,
        };
    }
    return {
        type: phase.type,
    };
}

export enum OpCodeReserved {
    send = 0x0,
    notSupported = 0xffffffff,
    notAllowed = 0xfffffffe,
}

export async function collectMetric<T extends Contract>(
    blockchain: Blockchain,
    ctx: MetricContext<T>,
    result: SendMessageResult,
) {
    const store = getMetricStore();
    if (!Array.isArray(store)) {
        return;
    }
    let state: StateMetric = { data: { cells: 0, bits: 0 }, code: { cells: 0, bits: 0 } };
    let codeHash: CodeHash | undefined;
    if (ctx.contract.init && ctx.contract.init.code && ctx.contract.init.data) {
        codeHash = `0x${ctx.contract.init.code.hash().toString('hex')}`;
        state = calcStateSize({ code: ctx.contract.init.code, data: ctx.contract.init.data });
    } else {
        const account = (await blockchain.getContract(ctx.contract.address)).accountState;
        if (account && account.type === 'active' && account.state.code && account.state.data) {
            codeHash = `0x${account.state.code.hash().toString('hex')}`;
            state = calcStateSize({ code: account.state.code, data: account.state.data });
        }
    }
    let testName;
    if ((globalThis as any)['expect']) {
        // eslint-disable-next-line no-undef
        testName = expect.getState().currentTestName;
    }
    let contractName: ContractName | undefined = ctx.contract.constructor.name;
    let methodName: ContractMethodName | undefined = ctx.methodName;

    for (const tx of result.transactions) {
        if (tx.description.type !== 'generic') continue;
        const receiver = tx.inMessage?.info.type;
        const body = tx.inMessage?.body ? tx.inMessage.body.beginParse() : undefined;
        let opCode: OpCode = '0x0';
        if (receiver === 'internal') {
            opCode = `0x${(body && body.remainingBits >= 32 ? body.preloadUint(32) : 0).toString(16)}`;
        }
        if (!methodName && Object.values(OpCodeReserved).includes(Number(opCode))) {
            methodName = OpCodeReserved[Number(opCode)];
        }
        const address = Address.parseRaw(`0:${tx.address.toString(16).padStart(64, '0')}`);

        const { computePhase, actionPhase } = tx.description;
        const action: ActionPhaseMetric | undefined = actionPhase
            ? {
                  success: actionPhase.success,
                  totalActions: actionPhase.totalActions,
                  skippedActions: actionPhase.skippedActions,
                  resultCode: actionPhase.resultCode,
                  totalActionFees: actionPhase.totalActions,
                  totalFwdFees: actionPhase.totalFwdFees ? Number(actionPhase.totalFwdFees) : undefined,
                  totalMessageSize: {
                      cells: Number(actionPhase.totalMessageSize.cells),
                      bits: Number(actionPhase.totalMessageSize.bits),
                  },
              }
            : undefined;
        const compute = calcComputePhase(computePhase);
        const metric: Metric = {
            testName,
            address: address.toString(),
            codeHash,
            contractName,
            methodName,
            receiver,
            opCode,
            execute: {
                compute,
                action,
            },
            message: {
                in: calcMessageSize(tx.inMessage),
                out: calcDictSize(tx.outMessages),
            },
            state,
        };
        store.push(metric);
        methodName = undefined;
        if (!address.equals(ctx.contract.address)) {
            contractName = ctx.contract.constructor.name;
        } else {
            contractName = undefined;
        }
    }
}
