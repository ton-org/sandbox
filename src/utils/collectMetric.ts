import { Address, beginCell, Cell, Contract } from '@ton/core';
import { Dictionary, DictionaryKeyTypes, TransactionComputePhase } from '@ton/core';
import { Blockchain, SendMessageResult } from '../blockchain/Blockchain';

export type MetricContext<T extends Contract> = {
    contract: T;
    methodName: string;
};

export type State = {
    code: Cell;
    data: Cell;
};

export type CellMetric = {
    cells: number;
    bits: number;
};

export type ComputePhaseMetric = {
    type: string;
    success?: boolean;
    gasUsed?: number;
    exitCode?: number;
    vmSteps?: number;
};

export type ActionPhaseMetric = {
    success?: boolean;
    totalActions?: number;
    skippedActions?: number;
    resultCode?: number;
    totalFwdFees?: number;
    totalActionFees?: number;
};

export type Metric = {
    testName?: string;
    address: string;
    contractName: string | null;
    methodName: string | null;
    opCode: number;
    computePhase: ComputePhaseMetric;
    actionPhase: ActionPhaseMetric;
    outMessages: CellMetric;
    state: CellMetric;
};

export type SnapshotMetric = {
    comment: string;
    createdAt: Date;
    items: Metric[];
};

const STORE_METRIC = Symbol.for('ton-sandbox-metric-store');

export function makeSnapshotMetric(comment: string, store: Metric[]): SnapshotMetric {
    const snapshot: SnapshotMetric = {
        comment,
        createdAt: new Date(),
        items: new Array<Metric>(),
    };
    const seen = new Set<string>();
    for (const metric of store) {
        const key = JSON.stringify(metric);
        if (seen.has(key)) continue;
        snapshot.items.push(metric);
        seen.add(key);
    }
    return snapshot;
}

export function getMetricStore(context: any = globalThis): Array<Metric> | undefined {
    return context[STORE_METRIC];
}

export function createMetricStore(context: any = globalThis): Array<Metric> {
    if (!Array.isArray(context[STORE_METRIC])) {
        context[STORE_METRIC] = new Array<Metric>();
    }
    return context[STORE_METRIC];
}

export function calcDictSize<K extends DictionaryKeyTypes, V>(dict: Dictionary<K, V>) {
    if (dict.size > 0) {
        return calcCellSize(beginCell().storeDict(dict).endCell().asSlice().loadRef());
    }
    return { cells: 0, bits: 0 };
}

export function calcCellSize(root: Cell, visited: Set<string> = new Set<string>()) {
    const hash = root.hash().toString('hex');
    if (visited.has(hash)) {
        return { cells: 0, bits: 0 };
    }
    visited.add(hash);
    let cells = 1;
    let bits = root.bits.length;
    for (const ref of root.refs) {
        const childRes = calcCellSize(ref, visited);
        cells += childRes.cells;
        bits += childRes.bits;
    }
    return { cells, bits };
}

export function calcStateSize(state: State): CellMetric {
    const codeSize = calcCellSize(state.code);
    const dataSize = calcCellSize(state.data);
    return {
        cells: codeSize.cells + dataSize.cells,
        bits: codeSize.bits + dataSize.bits,
    };
}

export function computePhase(phase: TransactionComputePhase): ComputePhaseMetric {
    if (phase.type === 'vm') {
        return {
            type: phase.type,
            success: phase.success,
            gasUsed: Number(phase.gasUsed),
            exitCode: phase.exitCode,
            vmSteps: phase.vmSteps,
        };
    }
    return {
        type: phase.type,
    };
}

export async function collectMetric<T extends Contract>(
    blockchain: Blockchain,
    ctx: MetricContext<T>,
    result: SendMessageResult,
) {
    const store = getMetricStore();
    if (!Array.isArray(store)) {
        return;
    }
    let state: CellMetric = { cells: 0, bits: 0 };
    if (ctx.contract.init && ctx.contract.init.code && ctx.contract.init.data) {
        state = calcStateSize({ code: ctx.contract.init.code, data: ctx.contract.init.data });
    } else {
        const account = (await blockchain.getContract(ctx.contract.address)).accountState;
        if (account && account.type === 'active' && account.state.code && account.state.data) {
            codeHash = `0x${account.state.code.hash().toString('hex')}`;
            state = calcStateSize({ code: account.state.code, data: account.state.data });
        }
    }
    let testName;
    if ((globalThis as any)['expect']) {
        testName = expect.getState().currentTestName;
    }
    let contractName: string | null = ctx.contract.constructor.name;
    let methodName: string | null = ctx.methodName;

    for (const tx of result.transactions) {
        if (tx.description.type !== 'generic') continue;
        const body = tx.inMessage?.info.type === 'internal' ? tx.inMessage?.body.beginParse() : undefined;
        const opCode = body && body.remainingBits >= 32 ? body.preloadUint(32) : 0;
        const address = Address.parse(`0:${tx.address.toString(16)}`);
        const metric: Metric = {
            testName,
            address: address.toString(),
            contractName,
            methodName,
            opCode,
            computePhase: computePhase(tx.description.computePhase),
            actionPhase: {
                success: tx.description.actionPhase?.success,
                totalActions: tx.description.actionPhase?.totalActions,
                skippedActions: tx.description.actionPhase?.skippedActions,
                resultCode: tx.description.actionPhase?.resultCode,
                totalActionFees: tx.description.actionPhase?.totalActions,
                totalFwdFees: Number(tx.description.actionPhase?.totalFwdFees ?? 0),
            },
            outMessages: calcDictSize(tx.outMessages),
            state,
        };
        store.push(metric);
        methodName = null;
        if (!address.equals(ctx.contract.address)) {
            contractName = ctx.contract.constructor.name;
        } else {
            contractName = null;
        }
    }
}
