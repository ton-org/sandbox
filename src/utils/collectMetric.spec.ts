import '@ton/test-utils';
import { beginCell, toNano } from '@ton/core';
import { Blockchain } from '../blockchain/Blockchain';
import { createMetricStore, getMetricStore, makeSnapshotMetric } from './collectMetric';

async function simpleCase() {
    const blockchain = await Blockchain.create();
    const [alice, bob] = await blockchain.createWallets(2);
    const res1 = await alice.send({
        to: bob.address,
        value: toNano('1'),
        body: beginCell().storeUint(3735943886, 32).endCell(),
    });
    expect(res1.transactions).toHaveTransaction({
        from: alice.address,
        to: bob.address,
        success: true,
    });
}

describe('collectMetric', () => {
    it('should not collect metric', async () => {
        await simpleCase();
        expect(getMetricStore()).toEqual(undefined);
    });

    it('should be collect metric', async () => {
        const store = createMetricStore();
        expect(store.length).toEqual(0);
        await simpleCase();
        expect(store.length).toEqual(2);
        const snapshot = makeSnapshotMetric('foo', store);
        expect(snapshot.comment).toEqual('foo');
        expect(snapshot.createdAt.getTime()).toBeLessThanOrEqual(new Date().getTime());
        expect(snapshot.items).toEqual([
            {
                testName: 'collectMetric should be collect metric',
                address: 'EQBiA46W-PQaaZZNFIDglnVknV9CR6J5hs81bSv70FwfNTrD',
                contractName: 'TreasuryContract',
                methodName: 'send',
                opCode: 0,
                computePhase: {
                    exitCode: 0,
                    gasUsed: 1937,
                    success: true,
                    type: 'vm',
                    vmSteps: 50,
                },
                actionPhase: {
                    resultCode: 0,
                    skippedActions: 0,
                    success: true,
                    totalActionFees: 1,
                    totalActions: 1,
                    totalFwdFees: 400000,
                },
                outMessages: {
                    bits: 744,
                    cells: 2,
                },
                state: {
                    bits: 714,
                    cells: 5,
                },
            },
            {
                testName: 'collectMetric should be collect metric',
                address: 'EQBc3CG3NOeF3wwkBM8zjXrsWUhjLuN45LobSkZHXCR0jhvg',
                contractName: null,
                methodName: null,
                opCode: 3735943886,
                computePhase: {
                    type: 'vm',
                    success: true,
                    exitCode: 0,
                    gasUsed: 309,
                    vmSteps: 5,
                },
                actionPhase: {
                    success: true,
                    resultCode: 0,
                    skippedActions: 0,
                    totalActionFees: 0,
                    totalActions: 0,
                    totalFwdFees: 0,
                },
                outMessages: {
                    bits: 0,
                    cells: 0,
                },
                state: {
                    bits: 714,
                    cells: 5,
                },
            },
        ]);
    });
});
