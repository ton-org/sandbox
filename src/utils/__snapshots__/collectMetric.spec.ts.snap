// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`collectMetric should be collect metric 1`] = `
[
  {
    "actionPhase": {
      "resultCode": 0,
      "skippedActions": 0,
      "success": true,
      "totalActionFees": 1,
      "totalActions": 1,
      "totalFwdFees": 400000,
    },
    "address": "EQBiA46W-PQaaZZNFIDglnVknV9CR6J5hs81bSv70FwfNTrD",
    "codeHash": "0xd992502b94ea96e7b34e5d62ffb0c6fc73d78b3e61f11f0848fb3a1eb1afc912",
    "computePhase": {
      "exitCode": 0,
      "gasUsed": 1937,
      "success": true,
      "type": "vm",
      "vmSteps": 50,
    },
    "contractName": "TreasuryContract",
    "inMessages": {
      "bits": 998,
      "cells": 3,
    },
    "methodName": "send",
    "opCode": "0x0",
    "outMessages": {
      "bits": 744,
      "cells": 2,
    },
    "receiver": "external-in",
    "state": {
      "code": {
        "bits": 458,
        "cells": 4,
      },
      "data": {
        "bits": 256,
        "cells": 1,
      },
    },
    "testName": "collectMetric should be collect metric",
  },
  {
    "actionPhase": {
      "resultCode": 0,
      "skippedActions": 0,
      "success": true,
      "totalActionFees": 0,
      "totalActions": 0,
      "totalFwdFees": 0,
    },
    "address": "EQBc3CG3NOeF3wwkBM8zjXrsWUhjLuN45LobSkZHXCR0jhvg",
    "codeHash": "0xd992502b94ea96e7b34e5d62ffb0c6fc73d78b3e61f11f0848fb3a1eb1afc912",
    "computePhase": {
      "exitCode": 0,
      "gasUsed": 309,
      "success": true,
      "type": "vm",
      "vmSteps": 5,
    },
    "contractName": "TreasuryContract",
    "inMessages": {
      "bits": 737,
      "cells": 1,
    },
    "methodName": "foo",
    "opCode": "0xdeadface",
    "outMessages": {
      "bits": 0,
      "cells": 0,
    },
    "receiver": "internal",
    "state": {
      "code": {
        "bits": 458,
        "cells": 4,
      },
      "data": {
        "bits": 256,
        "cells": 1,
      },
    },
    "testName": "collectMetric should be collect metric",
  },
  {
    "actionPhase": {
      "resultCode": 0,
      "skippedActions": 0,
      "success": true,
      "totalActionFees": 1,
      "totalActions": 1,
      "totalFwdFees": 400000,
    },
    "address": "EQBc3CG3NOeF3wwkBM8zjXrsWUhjLuN45LobSkZHXCR0jhvg",
    "codeHash": "0xd992502b94ea96e7b34e5d62ffb0c6fc73d78b3e61f11f0848fb3a1eb1afc912",
    "computePhase": {
      "exitCode": 0,
      "gasUsed": 1937,
      "success": true,
      "type": "vm",
      "vmSteps": 50,
    },
    "contractName": "TreasuryContract",
    "inMessages": {
      "bits": 998,
      "cells": 3,
    },
    "methodName": "send",
    "opCode": "0x0",
    "outMessages": {
      "bits": 744,
      "cells": 2,
    },
    "receiver": "external-in",
    "state": {
      "code": {
        "bits": 458,
        "cells": 4,
      },
      "data": {
        "bits": 256,
        "cells": 1,
      },
    },
    "testName": "collectMetric should be collect metric",
  },
  {
    "actionPhase": {
      "resultCode": 0,
      "skippedActions": 0,
      "success": true,
      "totalActionFees": 0,
      "totalActions": 0,
      "totalFwdFees": 0,
    },
    "address": "EQBiA46W-PQaaZZNFIDglnVknV9CR6J5hs81bSv70FwfNTrD",
    "codeHash": "0xd992502b94ea96e7b34e5d62ffb0c6fc73d78b3e61f11f0848fb3a1eb1afc912",
    "computePhase": {
      "exitCode": 0,
      "gasUsed": 309,
      "success": true,
      "type": "vm",
      "vmSteps": 5,
    },
    "contractName": "TreasuryContract",
    "inMessages": {
      "bits": 737,
      "cells": 1,
    },
    "methodName": "notSupported",
    "opCode": "0xffffffff",
    "outMessages": {
      "bits": 0,
      "cells": 0,
    },
    "receiver": "internal",
    "state": {
      "code": {
        "bits": 458,
        "cells": 4,
      },
      "data": {
        "bits": 256,
        "cells": 1,
      },
    },
    "testName": "collectMetric should be collect metric",
  },
]
`;
